_posts/2020-07-29-parallelized-sampling/parallelized-sampling.Rmd

```diff
@@ -0,0 +1,139 @@
+---
+title: "Parallelized sampling using exponential variates"
+description: > 
+  How can the seemingly iterative process of weighted sampling without replacement be transformed into something highly parallelizable? Turns out a well-known technique based on exponential variates accomplshes exactly that.
+author:
+  - name: Yitao Li
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+bibliography: bibliography.bib
+date: 07-29-2020
+categories:
+  - Probabilistic ML/DL
+  - Concepts
+output:
+  distill::distill_article:
+    self_contained: false
+preview: images/dice.jpg
+header-includes:
+   \usepackage{algorithm2e}
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+As part of my recent work to support weighted sampling of Spark dataframes in `sparklyr`, I embarked on a journey searching for algorithms that can perform weighted sampling, especially sampling without replacement, in efficient and scalable ways within a distributed cluster-computing framework such as Apache Spark.
+
+In the interest of brevity, "weighted sampling without replacement" shall be shortened into `SWoR` for the remainder of this blog post.
+
+In the following sections, I will explain and illustrate what `SWoR` means probability-wise, briefly outline some alternative solutions I have considered but was not completely satisfied with, and then deep-dive into exponential variates, a simple mathematical construct that made the ideal solution for this problem possible.
+
+## How it all started?
+
+My journey started from a [Github issue](https://github.com/sparklyr/sparklyr/issues/2592) inquiring about the possibility of supporting the equivalent of `dplyr::sample_frac(..., weight = <weight_column>)` for Spark dataframes in `sparklyr`. For example,
+
+```{r}
+dplyr::sample_frac(mtcars, 0.25, weight = gear, replace = FALSE)
+```
```

Comment:
I think is worth adding the output of `dplyr::sample_frac`, gives some context as to what we are trying to do.